src/csharp/Grpc.Core/Internal/NativeCallbackDispatcher.cs

```diff
@@ -0,0 +1,101 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Collections.Generic;
+using Grpc.Core.Logging;
+
+namespace Grpc.Core.Internal
+{
+    internal delegate int UniversalNativeCallback(IntPtr arg0, IntPtr arg1, IntPtr arg2, IntPtr arg3, IntPtr arg4, IntPtr arg5);
+
+    internal delegate int NativeCallbackDispatcherCallback(IntPtr tag, IntPtr arg0, IntPtr arg1, IntPtr arg2, IntPtr arg3, IntPtr arg4, IntPtr arg5);
+
+    internal class NativeCallbackDispatcher
+    {
+        static readonly ILogger Logger = GrpcEnvironment.Logger.ForType<NativeCallbackDispatcher>();
+        static readonly object staticLock = new object();
+
+        static NativeCallbackDispatcherCallback dispatcherCallback;
+
+        public static void Init(NativeMethods native)
+        {
+            lock (staticLock)
+            {
+                if (dispatcherCallback == null)
+                {
+                    dispatcherCallback = new NativeCallbackDispatcherCallback(HandleDispatcherCallback);
+                    native.grpcsharp_native_callback_dispatcher_init(dispatcherCallback);
+                }
+            }
+        }
+
+        public static NativeCallbackRegistration RegisterCallback(UniversalNativeCallback callback)
+        {
+            var gcHandle = GCHandle.Alloc(callback);
+            return new NativeCallbackRegistration(gcHandle);
+        }
+
+        private static UniversalNativeCallback GetCallback(IntPtr tag)
+        {
+            var gcHandle = GCHandle.FromIntPtr(tag);
+            return (UniversalNativeCallback) gcHandle.Target;
+        }
+
+        [MonoPInvokeCallback(typeof(NativeCallbackDispatcherCallback))]
+        private static int HandleDispatcherCallback(IntPtr tag, IntPtr arg0, IntPtr arg1, IntPtr arg2, IntPtr arg3, IntPtr arg4, IntPtr arg5)
+        {
+            try
+            {
+                var callback = GetCallback(tag);
+                return callback(arg0, arg1, arg2, arg3, arg4, arg5);
+            }
+            catch (Exception e)
+            {
+                // eat the exception, we must not throw when inside callback from native code.
+                Logger.Error(e, "Caught exception inside callback from native callback.");
+                return 0;
+            }
+        }
+    }
+
+    internal class NativeCallbackRegistration : IDisposable
+    {
+        readonly GCHandle handle;
+
+        public NativeCallbackRegistration(GCHandle handle)
+        {
+            this.handle = handle;
+        }
+
+        public IntPtr Tag => GCHandle.ToIntPtr(handle);
+
+        public void Dispose()
+        {
+            if (handle.IsAllocated)
```

Comment:
When would `handle.IsAllocated` ever be false? I could imagine that happening if we fail to initially allocate the handle, or double free. But both should be non-recoverable errors, no?

Reply:
It should be allowed to call Dispose() multiple times. (I think that's a IDisposable requirement).